File: app/src/main/java/com/voltbill/pro/domain/Gst.kt
package com.voltbill.pro.domain

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToLong

/** Rounds to 2 decimals the way accounting expects (half-up). */
fun money(v: Double): Double = (v * 100.0).roundToLong() / 100.0

/** ₹1,23,456.78 — Indian digit grouping. */
fun inr(v: Double): String {
    val neg = v < 0
    val a = abs(money(v))
    val whole = a.toLong()
    val paise = ((a - whole) * 100).roundToLong().toInt()
    val s = whole.toString()
    val grouped = if (s.length <= 3) s else {
        val last3 = s.substring(s.length - 3)
        var rest = s.substring(0, s.length - 3)
        val sb = StringBuilder()
        while (rest.length > 2) {
            sb.insert(0, "," + rest.substring(rest.length - 2))
            rest = rest.substring(0, rest.length - 2)
        }
        if (rest.isNotEmpty()) sb.insert(0, rest)
        "$sb,$last3"
    }
    return (if (neg) "-₹" else "₹") + grouped + "." + paise.toString().padStart(2, '0')
}

/** Result of costing a single invoice line. */
data class LineTax(
    val taxable: Double,
    val cgst: Double,
    val sgst: Double,
    val igst: Double,
    val total: Double
)

/**
 * Computes one line: qty × rate, less line discount %, then GST split.
 * Intra-state -> CGST+SGST (half each). Inter-state -> IGST.
 */
fun computeLine(
    qty: Double,
    rate: Double,
    discountPct: Double,
    gstRate: Double,
    interState: Boolean
): LineTax {
    val gross = qty * rate
    val taxable = money(gross - gross * (discountPct / 100.0))
    val tax = money(taxable * gstRate / 100.0)
    return if (interState) {
        LineTax(taxable, 0.0, 0.0, tax, money(taxable + tax))
    } else {
        val half = money(tax / 2.0)
        LineTax(taxable, half, half, 0.0, money(taxable + half * 2))
    }
}

data class InvoiceTotals(
    val subTotal: Double,
    val taxable: Double,
    val cgst: Double,
    val sgst: Double,
    val igst: Double,
    val exchange: Double,
    val roundOff: Double,
    val grandTotal: Double
) {
    val totalTax: Double get() = money(cgst + sgst + igst)
}

/** Aggregates lines, applies old-battery exchange deduction and round-off. */
fun computeTotals(lines: List<LineTax>, grossBeforeDiscount: Double, exchange: Double): InvoiceTotals {
    val taxable = money(lines.sumOf { it.taxable })
    val cgst = money(lines.sumOf { it.cgst })
    val sgst = money(lines.sumOf { it.sgst })
    val igst = money(lines.sumOf { it.igst })
    val beforeRound = money(taxable + cgst + sgst + igst - exchange)
    val rounded = Math.round(beforeRound).toDouble()
    return InvoiceTotals(
        subTotal = money(grossBeforeDiscount),
        taxable = taxable,
        cgst = cgst, sgst = sgst, igst = igst,
        exchange = money(exchange),
        roundOff = money(rounded - beforeRound),
        grandTotal = rounded
    )
}

/** "Rupees One Lakh Twenty Three Thousand Four Hundred Fifty Six and Seventy Eight Paise Only" */
fun amountInWords(amount: Double): String {
    val a = money(abs(amount))
    val rupees = a.toLong()
    val paise = ((a - rupees) * 100).roundToLong()
    val sb = StringBuilder("Rupees ").append(wordsFor(rupees))
    if (paise > 0) sb.append(" and ").append(wordsFor(paise)).append(" Paise")
    return sb.append(" Only").toString()
}

private val ones = arrayOf(
    "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
    "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen",
    "Eighteen", "Nineteen"
)
private val tens = arrayOf("", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety")

private fun twoDigits(n: Int): String = when {
    n == 0 -> ""
    n < 20 -> ones[n]
    else -> (tens[n / 10] + if (n % 10 != 0) " " + ones[n % 10] else "")
}

private fun threeDigits(n: Int): String {
    val h = n / 100
    val r = n % 100
    val sb = StringBuilder()
    if (h > 0) sb.append(ones[h]).append(" Hundred")
    if (r > 0) { if (h > 0) sb.append(" "); sb.append(twoDigits(r)) }
    return sb.toString()
}

/** Indian numbering: crore, lakh, thousand, hundred. */
fun wordsFor(value: Long): String {
    if (value == 0L) return "Zero"
    var n = value
    val parts = mutableListOf<String>()
    val crore = (n / 10_000_000L).toInt(); n %= 10_000_000L
    val lakh = (n / 100_000L).toInt(); n %= 100_000L
    val thousand = (n / 1_000L).toInt(); n %= 1_000L
    val rest = n.toInt()
    if (crore > 0) parts += threeDigits(crore) + " Crore"
    if (lakh > 0) parts += twoDigits(lakh) + " Lakh"
    if (thousand > 0) parts += twoDigits(thousand) + " Thousand"
    if (rest > 0) parts += threeDigits(rest)
    return parts.joinToString(" ")
}

// ---------- dates ----------

private val dmy = SimpleDateFormat("dd MMM yyyy", Locale.ENGLISH)
private val dmyShort = SimpleDateFormat("dd/MM/yy", Locale.ENGLISH)
private val full = SimpleDateFormat("dd MMM yyyy, hh:mm a", Locale.ENGLISH)

fun fmtDate(ts: Long): String = dmy.format(Date(ts))
fun fmtShort(ts: Long): String = dmyShort.format(Date(ts))
fun fmtDateTime(ts: Long): String = full.format(Date(ts))

fun addMonths(ts: Long, months: Int): Long = Calendar.getInstance().apply {
    timeInMillis = ts
    add(Calendar.MONTH, months)
}.timeInMillis

fun startOfDay(ts: Long = System.currentTimeMillis()): Long = Calendar.getInstance().apply {
    timeInMillis = ts
    set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
    set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
}.timeInMillis

fun endOfDay(ts: Long = System.currentTimeMillis()): Long = startOfDay(ts) + 86_399_999L

fun startOfMonth(ts: Long = System.currentTimeMillis()): Long = Calendar.getInstance().apply {
    timeInMillis = startOfDay(ts)
    set(Calendar.DAY_OF_MONTH, 1)
}.timeInMillis

fun endOfMonth(ts: Long = System.currentTimeMillis()): Long = Calendar.getInstance().apply {
    timeInMillis = startOfMonth(ts)
    add(Calendar.MONTH, 1)
}.timeInMillis - 1

/** Indian financial year start (April 1). */
fun startOfFinancialYear(ts: Long = System.currentTimeMillis()): Long {
    val c = Calendar.getInstance().apply { timeInMillis = startOfDay(ts) }
    val year = if (c.get(Calendar.MONTH) < Calendar.APRIL) c.get(Calendar.YEAR) - 1 else c.get(Calendar.YEAR)
    return Calendar.getInstance().apply {
        timeInMillis = 0
        set(year, Calendar.APRIL, 1, 0, 0, 0)
        set(Calendar.MILLISECOND, 0)
    }.timeInMillis
}

fun daysBetween(a: Long, b: Long): Int = ((b - a) / 86_400_000L).toInt()

/** VB/2025-26/0001 -> next serial in the current financial year. */
fun nextInvoiceNo(last: String?, prefix: String): String {
    val c = Calendar.getInstance()
    val y = if (c.get(Calendar.MONTH) < Calendar.APRIL) c.get(Calendar.YEAR) - 1 else c.get(Calendar.YEAR)
    val fy = "$y-${((y + 1) % 100).toString().padStart(2, '0')}"
    val seq = last?.substringAfterLast('/')?.toIntOrNull()?.plus(1) ?: 1
    return "$prefix/$fy/${seq.toString().padStart(4, '0')}"
}

File: crs-power-billing/src/lib/gst.js
/**
 * GST engine for C.R.S Power Solution.
 *
 * Pondicherry (Puducherry) — state code 34.
 * Intra-state supply  -> CGST + SGST (half the slab each)
 * Inter-state supply  -> IGST (full slab)
 *
 * Slabs used by the business:
 *   18% — inverters, UPS, solar controllers, stabilizers, accessories, services
 *   28% — lead-acid batteries (HSN 8507)
 *
 * All money is handled in paise internally to avoid floating point drift,
 * then returned as rupees rounded to 2 decimals.
 */

export const HOME_STATE_CODE = '34';
export const GST_SLABS = [18, 28];

/** Round half-up to 2 decimals (accounting convention). */
export function money(value) {
  const n = Number(value) || 0;
  return Math.round((n + Number.EPSILON) * 100) / 100;
}

/** Convert rupees to integer paise. */
function toPaise(rupees) {
  return Math.round((Number(rupees) || 0) * 100);
}

/** ₹1,23,456.78 — Indian digit grouping (lakh / crore). */
export function formatINR(value, { withSymbol = true } = {}) {
  const n = money(value);
  const negative = n < 0;
  const abs = Math.abs(n);
  const whole = Math.floor(abs);
  const paise = Math.round((abs - whole) * 100);

  const s = String(whole);
  let grouped;
  if (s.length <= 3) {
    grouped = s;
  } else {
    const last3 = s.slice(-3);
    let rest = s.slice(0, -3);
    const parts = [];
    while (rest.length > 2) {
      parts.unshift(rest.slice(-2));
      rest = rest.slice(0, -2);
    }
    if (rest.length) parts.unshift(rest);
    grouped = `${parts.join(',')},${last3}`;
  }

  const body = `${grouped}.${String(paise).padStart(2, '0')}`;
  return `${negative ? '-' : ''}${withSymbol ? '₹' : ''}${body}`;
}

/**
 * Cost a single invoice line.
 *
 * @param {object} line
 * @param {number} line.qty
 * @param {number} line.rate            price per unit, GST-exclusive
 * @param {number} [line.discountPct]   0-100
 * @param {number} line.gstRate         18 or 28
 * @param {boolean} interState
 */
export function computeLine(line, interState = false) {
  const qty = Number(line.qty) || 0;
  const rate = Number(line.rate) || 0;
  const discountPct = Number(line.discountPct) || 0;
  const gstRate = Number(line.gstRate) || 0;

  const grossPaise = toPaise(qty * rate);
  const discountPaise = Math.round((grossPaise * discountPct) / 100);
  const taxablePaise = grossPaise - discountPaise;
  const taxPaise = Math.round((taxablePaise * gstRate) / 100);

  let cgstPaise = 0;
  let sgstPaise = 0;
  let igstPaise = 0;

  if (interState) {
    igstPaise = taxPaise;
  } else {
    // split so the halves always sum exactly to the total tax
    cgstPaise = Math.floor(taxPaise / 2);
    sgstPaise = taxPaise - cgstPaise;
  }

  return {
    gross: money(grossPaise / 100),
    discount: money(discountPaise / 100),
    taxable: money(taxablePaise / 100),
    gstRate,
    cgst: money(cgstPaise / 100),
    sgst: money(sgstPaise / 100),
    igst: money(igstPaise / 100),
    tax: money(taxPaise / 100),
    total: money((taxablePaise + taxPaise) / 100),
  };
}

/**
 * Aggregate lines into invoice totals, applying an optional old-battery
 * exchange (buy-back) deduction and rupee round-off.
 */
export function computeInvoiceTotals(lines, { interState = false, exchange = 0 } = {}) {
  const computed = lines.map((l) => computeLine(l, interState));

  const sum = (key) => computed.reduce((acc, l) => acc + toPaise(l[key]), 0);

  const grossPaise = sum('gross');
  const discountPaise = sum('discount');
  const taxablePaise = sum('taxable');
  const cgstPaise = sum('cgst');
  const sgstPaise = sum('sgst');
  const igstPaise = sum('igst');
  const exchangePaise = toPaise(exchange);

  const beforeRoundPaise =
    taxablePaise + cgstPaise + sgstPaise + igstPaise - exchangePaise;

  const roundedPaise = Math.round(beforeRoundPaise / 100) * 100;
  const roundOffPaise = roundedPaise - beforeRoundPaise;

  return {
    lines: computed,
    subTotal: money(grossPaise / 100),
    discount: money(discountPaise / 100),
    taxable: money(taxablePaise / 100),
    cgst: money(cgstPaise / 100),
    sgst: money(sgstPaise / 100),
    igst: money(igstPaise / 100),
    totalTax: money((cgstPaise + sgstPaise + igstPaise) / 100),
    exchange: money(exchangePaise / 100),
    roundOff: money(roundOffPaise / 100),
    grandTotal: money(roundedPaise / 100),
    interState,
  };
}

/** Decide intra vs inter state from the customer's GSTIN / state code. */
export function isInterState(customerStateCode) {
  if (!customerStateCode) return false;
  return String(customerStateCode).trim() !== HOME_STATE_CODE;
}

/** HSN-wise tax summary for the GSTR-1 style report and the invoice footer. */
export function hsnSummary(lines, interState = false) {
  const map = new Map();
  for (const line of lines) {
    const c = computeLine(line, interState);
    const key = line.hsn || '—';
    const row = map.get(key) || {
      hsn: key,
      gstRate: line.gstRate,
      qty: 0,
      taxable: 0,
      cgst: 0,
      sgst: 0,
      igst: 0,
    };
    row.qty += Number(line.qty) || 0;
    row.taxable = money(row.taxable + c.taxable);
    row.cgst = money(row.cgst + c.cgst);
    row.sgst = money(row.sgst + c.sgst);
    row.igst = money(row.igst + c.igst);
    map.set(key, row);
  }
  return [...map.values()];
}

const ONES = ['', 'One', 'Two', 'Three', 'Four', 'Five', 'Six', 'Seven', 'Eight', 'Nine',
  'Ten', 'Eleven', 'Twelve', 'Thirteen', 'Fourteen', 'Fifteen', 'Sixteen', 'Seventeen',
  'Eighteen', 'Nineteen'];
const TENS = ['', '', 'Twenty', 'Thirty', 'Forty', 'Fifty', 'Sixty', 'Seventy', 'Eighty', 'Ninety'];

function twoDigits(n) {
  if (n === 0) return '';
  if (n < 20) return ONES[n];
  return TENS[Math.floor(n / 10)] + (n % 10 ? ` ${ONES[n % 10]}` : '');
}

function threeDigits(n) {
  const h = Math.floor(n / 100);
  const r = n % 100;
  const out = [];
  if (h) out.push(`${ONES[h]} Hundred`);
  if (r) out.push(twoDigits(r));
  return out.join(' ');
}

/** Indian numbering system: crore / lakh / thousand. */
export function wordsForNumber(value) {
  let n = Math.floor(Math.abs(Number(value) || 0));
  if (n === 0) return 'Zero';
  const parts = [];
  const crore = Math.floor(n / 10000000); n %= 10000000;
  const lakh = Math.floor(n / 100000); n %= 100000;
  const thousand = Math.floor(n / 1000); n %= 1000;
  if (crore) parts.push(`${threeDigits(crore)} Crore`);
  if (lakh) parts.push(`${twoDigits(lakh)} Lakh`);
  if (thousand) parts.push(`${twoDigits(thousand)} Thousand`);
  if (n) parts.push(threeDigits(n));
  return parts.join(' ');
}

/** "Rupees Fifty Three Thousand Twenty Two and Fifty Paise Only" */
export function amountInWords(value) {
  const n = money(Math.abs(value));
  const rupees = Math.floor(n);
  const paise = Math.round((n - rupees) * 100);
  let out = `Rupees ${wordsForNumber(rupees)}`;
  if (paise > 0) out += ` and ${wordsForNumber(paise)} Paise`;
  return `${out} Only`;
}

/**
 * Invoice numbering: CRS/2025-26/0001, restarting each Indian financial year
 * (which begins on 1 April).
 */
export function financialYear(date = new Date()) {
  const y = date.getFullYear();
  const startYear = date.getMonth() < 3 ? y - 1 : y; // Jan-Mar belong to the previous FY
  const endShort = String((startYear + 1) % 100).padStart(2, '0');
  return `${startYear}-${endShort}`;
}

export function nextInvoiceNo(lastInvoiceNo, prefix = 'CRS', date = new Date()) {
  const fy = financialYear(date);
  let seq = 1;
  if (lastInvoiceNo) {
    const parts = String(lastInvoiceNo).split('/');
    const lastFy = parts[1];
    const lastSeq = parseInt(parts[2], 10);
    if (lastFy === fy && Number.isFinite(lastSeq)) seq = lastSeq + 1;
  }
  return `${prefix}/${fy}/${String(seq).padStart(4, '0')}`;
}
